### DISCUSSION TITLE
Wider applications of {epiparameter} extraction

### DISCUSSION THREAD

**Discussion by adamkucharski** [epiverse-trace/epiverse-trace.github.io#319] [Blog post ideas]
I'm finding the functions in {epiparameter} very helpful for defining plausible distributions based on summary statistics in the literature, e.g. for [a prior in H5N1 cluster analysis](https://github.com/epiverse-trace/howto/blob/h5_example/analyses/quantify_transmission/reproduction_number_cluster_size.qmd) or @sarahdicksonuoft's work that required distribution of measles vaccine coverage.

Could be nice to have a post outlining other use cases if anyone has specific examples of this having been done manually in the past (e.g. papers converting literature estimate with 95% CrI into a parameteric prior).

**Comment by TimTaylor** [comment]
That looks good! Very out of the loop with things at the moment so maybe this has already been discussed but I wonder if densities/distributions/quantiles/random generation functions could be an output from epiparameter?
E.g.
```R
# Get prior for R based on Aditama et al, PLOS ONE, 2012
get_prior <- extract_param(
  type = "percentiles",
  values = c(0.009, 0.315),
  distribution = "gamma",
  percentiles = c(0.025,0.975)
)

h5_prior_r <- function(x){dgamma(x,shape = get_prior[["shape"]], scale = get_prior[["scale"]])}
```
could become
```
h5_prior_r <- get_prior[["dgamma"]]
```

I'm not sure though. Maybe this would be too opaque?

**Reply by adamkucharski** [reply]
Would be interested to hear @joshwlambert's thoughts, as reworking of parameter definitions in {epiparameter} could make this easier?

**Reply by TimTaylor** [reply]
Perhaps it could utilise https://alexpghayes.github.io/distributions3/index.html as well. I quite like the route that seems to be going

**Reply by joshwlambert** [reply]
@TimTaylor were you thinking of something like this:

``` r
library(epiparameter)
measles_R_prior <- epiparameter(
  disease = "measles",
  epi_dist = "R prior", 
  prob_distribution = "gamma", 
  summary_stats = create_summary_stats(
    quantiles = c("2.5" = 0.009, "97.5" = 0.315)
  )
)
#> Citation cannot be created as author, year, journal or title is missing
#> Stochastic numerical optimisation used. 
#> Rerun function multiple times to check global optimum is found
measles_R_prior
#> Disease: measles
#> Pathogen: NA
#> Epi Distribution: r prior
#> Study: (????). "No citation."
#> Distribution: gamma
#> Parameters:
#>   shape: 1.641
#>   scale: 0.064
density(measles_R_prior, at = 5)
#> [1] 2.714256e-32
cdf(measles_R_prior, q = 0.5)
#> [1] 0.9982221
quantile(measles_R_prior, p = 0.5)
#> [1] 0.08438875
generate(measles_R_prior, times = 10)
#>  [1] 0.006428562 0.062292303 0.032599392 0.122986171 0.019343796 0.082148907
#>  [7] 0.087469412 0.124865129 0.062132800 0.269208190
```

<sup>Created on 2024-09-27 with [reprex v2.1.0](https://reprex.tidyverse.org)</sup>

**Reply by joshwlambert** [reply]
In writing this I noticed a difference in how percentiles are supplied to `extract_param()` ([0, 1]) and `epiparameter()` ([0, 100]), so we should standardise this to not confuse users. Also standardising the use of either `percentile` or `quantile` which are often used interchangably would be more consistent.

**Reply by TimTaylor** [reply]
Yes. If `extract_param` also returned an object that you can call density/cdf/quantile on I ***think*** that would feel nice. Again, sorry if these things have been discussed before (I'm quite out of the loop at the mo).

**Reply by joshwlambert** [reply]
I'm not sure. My immediate thought is it is overcomplicating the output (which is currently a `numeric` vector) and is somewhat duplicating the functionality of `<epiparameter>` by having distribution methods. 

We currently depend on {distributional} so we can utilise their classes and methods for this, so it wouldn't add dependencies unless there was clear reasoning to go with {distributions3}. 

I'm happy to be convinced of making the change. One possible benefit with some upcoming changes in {epiparameter} would be to do something like this:

```r
dist <- extract_param(...)
epiparameter(
  disease = "<disease>",
  epi_dist = "<epi_dist>",
  prob_distribution = dist
)
```

**Reply by TimTaylor** [reply]
~Yeah - think you're right. I need to be more familiar with what the package. I misunderstood the example.~

**Reply by TimTaylor** [reply]
@joshwlambert - I'm still pondering. It feels like `extract_param()` is almost an `extract_distribution()` type thing so you could have the case you have above

```R
dist <- extract_param(...)
epiparameter(
  disease = "<disease>",
  epi_dist = "<epi_dist>",
  prob_distribution = dist
)
```
and also

```R
dist <- extract_param(...)
density(dist, ...)
cdf(dist, ...)
```
etc, etc

No preference as to either package (just liked the initial feel when I saw distributions3). Was unaware of distributional. Looking at the available options it's only a matter of time until some one makes a distributions7 (with s7) :sweat_smile:

**Reply by joshwlambert** [reply]
Thanks for the input @TimTaylor, I'm still unsure. Please could you open this as an issue in {epiparameter} and we can discuss the details there. 

For me the decider would be if there is a use case where `extract_param()` would need to be used directly used as a distribution.

Some other options:
- add an `extract_distribution()` function do what you're suggesting
- add an argument to `extract_param()` called something like `as_distribution` or `vector` which could take a logical and then the output could be changed accordingly.

**Reply by TimTaylor** [reply]
> - add an extract_distribution() function do what you're suggesting

This would probably be the best approach if desired/needed.

**Comment by joshwlambert** [comment]
@adamkucharski thanks for posting, I like the idea of having a post showcasing examples of where the conversion and extraction functions can help users for specific tasks. We'd need a list of examples and it might be a fair amount of work to search the literature for cases. I'd be happy to collaborate on the post.

**Reply by adamkucharski** [reply]
Good point – potentially better to keep this post as a placeholder for now, so can add examples as they crop up (and solicit from others), rather than focusing on getting out ASAP.

**Comment by adamkucharski** [comment]
In case useful, also highlighting a similar tool in python: https://preliz.readthedocs.io/en/latest/examples/param_space_1d_examples.html#from-intervals-to-maximum-entropy-distributions
